templates/settings.py.erb

```diff
@@ -17,4 +17,6 @@ REDIS_HOST = "localhost"
 REDIS_PORT = "<%= scope['redis::port'] %>"
 REDIS_DB = <%= scope['pulpcore::redis_db'] %>
 WORKING_DIRECTORY = "<%= scope['pulpcore::cache_dir'] %>"
-REMOTE_USER_ENVIRON_NAME = 'HTTP_REMOTE_USER'
+<% if defined?(scope['pulpcore::django_remote_user']) -%>
```

Comment:
Latest issue:

```
Failures:

  1) pulpcore on centos-7-x86_64 is expected to contain Concat::Fragment[base] with content !~ /REMOTE_USER_ENVIRON_NAME/

     Failure/Error: it { is_expected.to contain_concat__fragment('base').without_content(/REMOTE_USER_ENVIRON_NAME/) }

     

       expected that the catalogue would contain Concat::Fragment[base] with content not set to /REMOTE_USER_ENVIRON_NAME/ but it is set to "CONTENT_HOST = \"foo.example.com\"\nCONTENT_ORIGIN = \"http://foo.example.com\"\nPULP2_MONGODB = {\"name\": \"pulp_database\", \"seeds\": \"localhost:27017\"}\nSECRET_KEY = \"qoXnTdTofHouJcuARF4XZruhoEtqzhwk\"\nDATABASES = {\n    'default': {\n        'ENGINE': 'django.db.backends.postgresql',\n        'NAME': 'pulpcore',\n        'USER': 'pulp',\n        'PASSWORD': 'votFojScBLa3jbR33byoGMbMGQPabg9M',\n        'HOST': '127.0.0.1',\n        'PORT': '5432',\n    }\n}\nMEDIA_ROOT = \"/var/lib/pulp/docroot\"\nREDIS_HOST = \"localhost\"\nREDIS_PORT = \"6379\"\nREDIS_DB = 8\nWORKING_DIRECTORY = \"/var/lib/pulp/tmp\"\nREMOTE_USER_ENVIRON_NAME = ''\n"
```

This suggests that the template rendered to include `REMOTE_USER_ENVIRONMENT_NAME = ''` when the whole line should have been omitted. This implies that something is wrong with our approach of using `if defined?(scope['pulpcore::django_remote_user'])`

Reply:
I think in the case where `$pulpcore::django_remote_user = undef` then `scope['pulpcore::django_remote_user']` is getting evaluated as `''` so we get `if defined?('')` which evalues as `true` and renders the block.

I think we need something like `if defined?(@pulpcore::django_remote_user)`

Reply:
Per https://docs.huihoo.com/puppet/guides/templating.html we have a few options.

`if scope['pulpcore::django_remote_user'] != ''` seems straightforward, as I can't see an use case where we ever want to write the line in the config without actually defining the setting.

Another option I think is adding `$django_remote_user = $pulpcore::django_remote_user` inside `manifests/config.pp` so that the template could do `if defined?(@django_remote_user)` without going out of scope.

Will get with https://github.com/synkd in the morning and try this.

Reply:
Second approach wouldn't work: https://puppet.com/docs/puppet/5.5/lang_template_erb.html

Puppet's `undef` translates to `nil:nilClass` in Ruby. This explains an undefined method defined? for class nil:nilClass that we saw previously.

So I think `if scope['pulpcore::django_remote_user'] != ''` is the correct approach.

Reply:
The variable is always defined so `defined?` always evaluates to true. I wonder if we should switch the template to [epp](https://puppet.com/docs/puppet/latest/lang_template_epp.html) and use Puppet logic. Then we can write a check like `<% if $remote_user_environ_name =~ String[1] -%>`

Reply:
Btw, did you check whether just `if scope['pulpcore::django_remote_user']` works?

Reply:
I thought we had tried that yesterday but apparently not -- this time it works.